lib/logfmt/encoder.ex

```diff
@@ -14,40 +51,38 @@ defmodule Logfmt.Encoder do
   @doc """
   See [`Logfmt.encode`](/logfmt/Logfmt.html#encode/1).
   """
-  @spec encode(Dict.t) :: String.t
-  def encode(list) do
-    Enum.reduce list, "", &encode_pair/2
-  end
-
-  @spec encode_pair({Dict.key, String.t | boolean | number}, String.t) :: String.t
-  defp encode_pair({key, value}, "") when is_binary(value) do
-    if String.contains?(value, "=") do
-      ~s(#{key}="#{encode_value(value)}")
-    else
-      ~s(#{key}=#{encode_value(value)})
+  @spec encode(Dict.t, options :: Keyword.t) :: String.t
+  def encode(list, options \\ []) do
+    result = Enum.reduce(list, [], &encode_pair/2)
+    case options[:output] do
+      # in case you wish to get back an iolist
+      :iolist -> result
+      # otherwise default to returning a binary
+      _ -> Enum.join(result)
     end
   end
 
-  defp encode_pair({key, value}, "") do
-    ~s(#{key}=#{encode_value(value)})
+  @spec encode_value(value :: term) :: String.t
+  defp encode_value(value) do
+    str = Logfmt.ValueEncoder.encode(value)
+    cond do
+      # originally I wanted to check if the string contained both ' and "
+      # the quoting would require escaping those characters first
+      String.contains?(str, " ") or String.contains?(str, "=") ->
+        ["\"", str, "\""]
+      true -> str
+    end
   end
 
-  @spec encode_pair({Dict.key, String.t | boolean | number}, String.t) :: String.t
-  defp encode_pair({key, value}, line) do
-    "#{line} #{encode_pair({key, value}, "")}"
+  defp encode_pair({key, value}, []) do
+    # this is called for the first item in the list, since the accumulator
+    # would be empty at that point.
+    [[encode_value(key), "=", encode_value(value)]]
```

Comment:
This isn't actually necessary—it results in something that looks like this:

```elixir
iex> Logfmt.Encoder.encode([foo: 1, bar: 2], output: :iolist)
[[["foo", "=", "1"]], " ", ["bar" ,"=", "2"]]
```

Unwrap the list, and you'll correctly get:

```elixir
iex> Logfmt.Encoder.encode([foo: 1, bar: 2], output: :iolist)
[["foo", "=", "1"], " ", ["bar" ,"=", "2"]]
```

The `output: :iolist` should have a test case, too.

Reply:
Ah, yes originally I was using the list join `|`, hence the double wrapped list, but I threw out the reverse code and never changed it back.